### ISSUE TITLE
Prepared statement deallocation

### ISSUE THREAD

**Issue by jakajancar** [mauricio/postgresql-async#111]
Reading the README, it seems we have two options:
- don't use prepared statements, and manually escape the parameter values, or
- use prepared statements and keep them alive forever

I would like to prepare a statement, execute it and deallocate it.

**Comment by farico** [comment]
Any update on this ticket? I also need a way to release statements. I have a dynamic query which does `INSERT INTO table (a, b, c) VALUES (1, 2, "user input"), (3, 5, "sql_injection_vuln"), ... , N ON DUPLICATE KEY UPDATE SET c = VALUES(c);` my values list is always dynamic and I believe a lot of prepared statements will use up MySQL memory. How to avoid this (and SQL injection)?

**Comment by mauricio** [comment]
@farico still on the plans, but contributions for it as welcome. Just providing a method that compiles and returns the compiled prepared statement would be good enough I think.

**Comment by dylex** [comment]
I wrote most of a `releasePreparedStatement` (really just `parsedStatements.remove(query)` plus `Close`) for postgres, but don't know enough about mysql to help there.  I can add some tests and send a PR for that part if it's useful, though.

**Comment by lpfeup** [comment]
Any update on this?

**Comment by mauricio** [comment]
Yes, there's a branch with work on it, I need to finish the MySQL part: https://github.com/mauricio/postgresql-async/tree/closing-prepared-statements

**Comment by KyleU** [comment]
We're getting demolished by this issue (MySQL). Our app creates a near-infinite set of potential queries, but we need placeholders and strongly typed parameters. Is there a safe way to inject placeholders without hitting the stored procedure cache? Something like `forPlaceholders(sql: String, values: Any*): String`. If we can't get past this, I'll  have to waste a week converting to JDBC, so I'm very motivated to get this resolved. Does the fix described in https://github.com/jilen/postgresql-async/blob/0.2.20-DRIP/mysql-async/src/main/scala/com/github/mauricio/async/db/mysql/codec/MySQLConnectionHandler.scala#L63 fix the unbounded statement cache?

**Comment by vitalif** [comment]
Hi. Another issue caused by this is mixing DDL and DML. PostgreSQL gives "Cached plan must not change result type" error when you try to use the same prepared statement (select *) with a modified table (for example, when a column has been added). The error itself is idiotic (I think DBMS must recompile the statement automatically in such case), but to workaround it we also need to deallocate prepared queries...

**Comment by oshai** [comment]
I merged the fix in jasync fork if someone interested: https://github.com/jasync-sql/jasync-sql/issues/82

**Comment by eugenemiretsky** [comment]
@KyleU  How did you solve this issue? We are facing the same problem.

**Comment by kyleu** [comment]
I switched to JDBC :|
